### DISCUSSION TITLE
Generate UI with Rust

### DISCUSSION THREAD

**Discussion by zmerp** [slint-ui/slint#593] [Q&A]
Is it possible to append/insert widgets on the Rust side?

**Comment by ogoffart** [comment]
Hi,
What's the use case? What exactly are you trying to do?

**Reply by zmerp** [reply]
I want a UI for editing settings, which should be generated from a json schema + json data. The settings form a tree arbitrarily deep. I don't want to redefine all settings structures in .60 because I already have them in Rust (which generate the json in the first place using a macro I wrote). 
This is the schema generator: https://github.com/alvr-org/ALVR/tree/master/alvr/settings-schema and this is the UI generator i'm currently using https://github.com/alvr-org/ALVR/blob/master/alvr/dashboard/js/app/settings.js but I want to move away from html/javascript bacause I can't use a webview in the platform I'm trying to port the project to.

**Reply by ogoffart** [reply]
The way i'd do it is to have something like 

```qml
VerticalLayout {
  for setting in settings : Rectangle {
     if setting.is-string-setting : StringSetting { data: setting.string-data }   
     if setting.is-bool-setting : BoolSetting { data: setting.bool-data }
  }
}
```
And i would flatten the tree in a `settings` model. There is an example of tree view in cargo-ui.

Not sure if that's really what you want.

Otherwise you could generate .60 code and use the sixtyfps-interpreter API.

**Reply by zmerp** [reply]
Thanks. using sixtyfps_interpreter is workable but a bit cumbersome. It would be great if built-in components like Rectangle and Text, and insert/append methods were available in the Rust API, even as part of sixtyfps_interpreter.
